internal/api/promote_subscribers_v1alpha1.go

```diff
@@ -0,0 +1,90 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"connectrpc.com/connect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubev1alpha1 "github.com/akuity/kargo/api/v1alpha1"
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	"github.com/akuity/kargo/internal/kargo"
+	"github.com/akuity/kargo/internal/logging"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+	v1alpha1 "github.com/akuity/kargo/pkg/api/v1alpha1"
+)
+
+func (s *server) PromoteSubscribers(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.PromoteSubscribersRequest],
+) (*connect.Response[svcv1alpha1.PromoteSubscribersResponse], error) {
+	if err := validateProjectAndStageNonEmpty(req.Msg.GetProject(), req.Msg.GetStage()); err != nil {
+		return nil, err
+	}
+	if err := s.validateProject(ctx, req.Msg.GetProject()); err != nil {
+		return nil, err
+	}
+	stage, err := getStage(ctx, s.client, req.Msg.GetProject(), req.Msg.GetStage())
+	if err != nil {
+		return nil, err
+	}
+	if err = validateFreightExists(req.Msg.GetFreight(), stage.Status.History); err != nil {
+		return nil, err
+	}
```

Comment:
For a _regular_ promotion, the target state must be among the _available_ states. Only healthy states from upstream flow downstream, so for a _regular_ promotion, the target state is always a healthy one.

For this case where we're pushing state instead of pulling, I wonder if there should be some check to avoid promotion of a state that's known to be no good.

Reply:
I didn't actually realize there might be freight in freight history that was not considered healthy. I think the suggestion makes sense.

Reply:
Added health check.